nemo/src/program_analysis/analysis.rs

```diff
@@ -660,6 +660,92 @@ impl ChaseProgram {
                 })
                 .collect::<HashMap<_, _>>();
 
+            let mut fact_and_constant_term_decls: HashMap<Identifier, Vec<Option<PrimitiveType>>> =
+                HashMap::new();
+            for fact in self.facts() {
+                let new_types: Vec<Option<PrimitiveType>> = fact
+                    .terms()
+                    .iter()
+                    .map(|c| c.primitive_type())
+                    .map(Some)
+                    .collect();
+
+                fact_and_constant_term_decls
+                    .entry(fact.predicate())
+                    .and_modify(|types| {
+                        *types = types
+                            .iter()
+                            .zip(new_types.iter())
+                            .map(|(a, b)| {
+                                a.as_ref()
+                                    .map(|a| b.as_ref().map(|b| a.max_type(b)).unwrap_or(*a))
+                                    .or(*b)
+                            })
+                            .collect();
+                    })
+                    .or_insert(new_types);
```

Comment:
I'm not following. Where does `types` vs. `new_types` come from? Why do we take the max?